ipac-ng/fetchipac.8
.\" Process this file with
.\" groff -man -Tascii foo.1
.\" $Id: fetchipac.8,v 1.15 2011/09/25 23:21:09 smsoft Exp $
.\"
.\" =()<.TH FETCHIPAC 8 "@<MANTIME>@" Linux "IPAC-NG">()=
.TH FETCHIPAC 8 "SEPTEMBER 2011" Linux "IPAC-NG"
.SH NAME
fetchipac \- ip accounting kernel data and database handler
.SH SYNOPSIS
.B fetchipac
[
.B -b|--batch
]
[
.B -d|--directory DIR
]
[
.B -h|--help
]
[
.B -m|--machine-output-format
]
[
.B -r|--records
]
[
.B -R|--list-raw
]
[
.B -s|--storage-method METHOD
]
[
.B -S|--start
]
[
.B -t|--timestamps [START][,END]
]
[
.B -v|--verbose
]
.SH DESCRIPTION
.B fetchipac
is part of the
.B ipac-ng
linux ip accounting package.

.B fetchipac
reads the kernel ip counters and stores this information in a database.
It also
provides access to the database for other programs, mainly for
.BR ipacsum (8).

If you are an average user of ipac, you don't want to know about 
most of the things described here. You will want to run fetchipac
regulary from cron, probably without any options. You will possibly
want to use the options --directory and --storage-method. The rest
is mostly for internal use by ipac and for debugging the database.

.B fetchipac
operates in different modes and does fairly unrelated things. The only common
thing it does is that it accesses a database. The database can be of different
types which are called
.I storage 
.IR methods .
The storage methods provided are:
.\" =()<.I @<STORAGEMETHODS>@.>()=
.I plain-file .
The default storage method used is
.\" =()<.IR @<DEFAULT_STORAGE>@.>()=
.IR plain-file.
.B fetchipac
uses a directory to store data. The default 
directory for that is
.\" =()<.IR @<ACCTDIR>@ .>()=
.IR /var/lib/ipac .
If the storage method does not make use of this directory, for example, if
it stores the data on a remote host, the directory is still
used for the spool file (see below).
(All these defaults are determined at build time and this man page is
updated automatically.)

.SH NORMAL OPERATION
When
.B fetchipac
is called without any arguments, it fetches kernel ip accounting data.

To do this, it first runs through a lock mechanism to prevent multiple
instances of
.B fetchipac
to run at the same time. Then, it reads the ipac rule file,
.\" =()<.IR @<RUNFILE>@ ,>()=
.IR /var/run/ipac.rules ,
which maps kernel ip accounting data to names. Next,
it reads the kernel ip accounting data from the appropriate file in
/proc/net, which is
.\" =()<.I @<IPCHAINS_PROC_C>@>()=
.I /proc/net/ip_fwchains
if you have an ipchains system (kernel 2.1.* or newer) 
.B fetchipac
opens this file in read/write mode; this makes the kernel set all counters
to zero in the moment the file is opened and we can be sure to
count every byte only once (see 
.B BUGS
section. In case of kernel 2.4.* it uses kernel's internal structures to
read byte and packet counters.

After reading and closing those files, 
.B fetchipac
creates a new record in its database using the default storage method.
.SH SPOOL MECHANISM
There is a
.I spool mechanism 
which allows 
.B fetchiapc
to spool ip accounting data in a file when the database cannot be opened.
For example, if the storage method stores data on a remote host, the remote host
might be down or unreachable. Another example is the gdbm storage method;
if some other process accesses a gdbm file, it is locked and can't be written 
to.
If 
.B fetchipac
could not spool the data, it would be lost in either case.

If the database can't be opened, the data collected from the kernel is
put into a spool file. The spool file resides in the data directory and
its name is
.IR spool .
If there is already data in this spool file, the new records are appended.
The data is stored in this file in the same format as described under the
batch mode command ADD (see below).

Whenever 
.B fetchipac
runs, it checks if the spool file exists. If so, and if the database can
be opened, the contents of the spool file is copied into the database and
the spool file is removed.

.SH COMMON OPTIONS
.TP
.B -d|--directory DIR
Use a different data directory DIR instead of the default,
.\" =()<.IR @<ACCTDIR>@ ,>()=
.IR /var/lib/ipac ,
for the database (if the 
storage method uses a directory to store data) and the spool file.
.TP
.B -h|--help
Print a help message and exit.
.TP
.B -s|--storage-method METHOD
Use METHOD as the storage method. METHOD can
be one of the compiled in storage methods. Run 
.I fetchipac --help
to see which storage methods exactly have been compiled in.
.TP
.B -S|--start
initial accounting setup: insert accounting rules and chains according to
your 
.\" =()<.I @<RULESFILE>@>()=
.I /etc/ipac-ng/ipac.conf
.RS

.SH OPTIONS USED INTERNALLY AND FOR DEBUGGING
.TP
.B -b|--batch
Switch to batch mode. In this mode, 
.B fetchipac
does not fetch ip accounting data from the kernel. Instead, it displays a
prompt, "> ", on standard output and then expects commands.
.B fetchipac
exists when the command QUIT was entered or at end of file on stdin. See
below, section BATCH MODE, for a description of the commands.
.TP
.B -m|--machine-output-format
When using the --record or --timestamps options,
.B fetchipac
normally prints the data in a nice human readable format. When using
this option, the data will be formatted in a way that it is 
easily parsable by a program. This is used by 
.BR ipacsum (8).

When used with --record, the output generated will be in the
format 
.B fetchipac
expects in batch mode with the ADD command.

When used with --timestamps, the output generated will be
in the format expected on standard input with the
--record option.
.TP
.B -r|--records
Switch to record retrieve mode. In this mode,
.B fetchipac
does not read the kernel ip accounting data from the kernel. 
Instead, it reads lines from standard input and exits on
end of file on stdin. The lines are interpreted as decimal numbers
which represent timestamps. (The lines may begin with a
single character '+', '-' or '*', which is discarded. The
number may also be preceded by white spaces.)

For every timestamp read, 
.B fetchipac
tries to retrieve the record from the database. If such a
record is not found, it prints "ERROR". If it is found,
the record is displayed in a human readable format.
(See also option --machine-output-format.)
.TP
.B -R|--list-raw
List rules and chains to be inserted into firewall. Don't actually insert
them, just list.
.TP
.B -t|--timestamps [START][,END]
Print all timestamps for which there are records in the

ipac-ng/CHANGELOG
# $Id: CHANGELOG,v 1.38 2011/11/05 01:25:58 smsoft Exp $
#
# description:
#   (name1/name2) ..... contributed software just got merged
#   (name1 + name2) ... software got merged and enhanced by developer
#      name1 = developer
#      name2 = contributor
#   (name1,name2) ..... work by the named developer
#                       no name means most likely that Al Zakharov did it
#
# current developers:
#   kaiser13/az ......... Al Zakharov <kaiser13 at users dot sourceforge dot net>
#   friedl/fl ........... Friedrich Lobenstock <fl at fl dot priv dot at>
#   mpoos ............... Mark Spieth <mpoos at users dot sourceforge dot net>
#   sm ...................Sergey Merzlikin <sm at smsoft dot ru>
#
# contributors:
#   Thomas Zehetbauer <thomasz at hostmaster dot org>
#   Simon Hausman <simon at lst dot de>
#   Denis O.Philippov <info at linuxcenter dot ru>
#   Jason Hudgins <thanatos_101 at yahoo dot com>
#   Mark Wormgoor.com <mark at wormgoor dot com>
#   xavier <list.ipac-ng-developer at natch dot dyndns dot org>
#   Afanasyev Nikolay <ikto at users dot sourceforge dot net>
#   Robert Evert <crimofbai at web dot de>
#

known problem(s):
    * ipactest script reports errors despite everything is 
      working ok - no fix yet
    * duplicate tickmarks in generated images can happen, this 
      is caused by rounding - no fix yet
    * with storage backend "gdbm" DO NOT call fetchipac twice in the 
      same second or your database will be destroyed - no fix, other than
      removing support for gdbm, yet
    * incorrect parsing of port ranges (port:port) in rules file.
    * if source/destination in rules file is specified as hostname,
      and this hostname resolves to multiple ip addresses, the rule
      is not accepted.
    * error message doesn't contain rule name if rule parsing error occurs

1.34.2
- fixed incompatibility with iptables 1.4.11 - 1.4.12.1 (sm)
- improved configure/build scripts. Now both dynamic (with iptables dynamic
  libraries) and static (with iptables source) compilation is compatible
  with iptables 1.4.3 - 1.4.12.1 (sm)
- updated documentation (sm)

1.34.1
- fixed incompatibility with some linux distributions which use iptables 1.4.10
  and later. Confirmed ability to work with iptables 1.4.3 - 1.4.11.1 (with shared
  iptables libraries, libxtables.so, libiptc.so and libip4tc.so). Static compilation
  works only with iptables version up to 1.4.9.1, and not in all linux distributions 
  (due to patched header files) (sm)

1.34
- ported to iptables 1.4.9.1, works with iptables 1.4.3 - 1.4.9.1, and maybe later (sm)
- removed a lot of unused code from iptables.c (sm)
- removed xtables.c, libiptc.c, libip4tc.c and corresponding header files,
  but instead added dependency on iptables libraries: libxtables.so and libiptc.so (sm)
- added option to configure script: "--with-iptables-src=" which allows to compile
  fetchipac statically with iptables source code (as it was in previous versions) (sm)
- added ipacsumdaily/monthly perl scripts to contrib directory (sm)
- updated documentation (sm)

1.33
- ported to iptables 1.4.2 (mdw21)

1.32

ipac-ng/fetchipac.c
/*
 *
 * $Id: fetchipac.c,v 1.22 2009/08/02 13:49:41 mdw21 Exp $
 *
 * Fetch IP accounting stats
 * Copyright (C) 1997 - 2000 Moritz Both
 *           (C) 2001 - 2003 Al Zakharov
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 * The author can be reached via email: moritz@daneben.de, or by
 * snail mail: Moritz Both, Im Moore 26, 30167 Hannover,
 *             Germany. Phone: +49-511-1610129
 *
 * Current developer and maintainer can be reached via kaiser13@mail2000.ru
 *
 */

#include <string.h>
#include <fnmatch.h>
#include <stdio.h>
#include <sys/stat.h>
#include <sys/wait.h>
#include <dirent.h>
#include <errno.h>
#include <search.h>
#include <time.h>
#include <stdlib.h>
#include <unistd.h>
#include <getopt.h>
#include <signal.h>
#include <fcntl.h>
#include "config.h"
#include <assert.h>
#include "ipac.h"

int		dropzero = 0;

int		verbose = 0;
/** name of the program */
const char 	*me;

/** the directory with the database.
 *  At least for plain-file and gdbm storage backend - and in any case for
 *  the spool file.
 */
const char	*datadir=ACCTDIR;

char		*spoolfile;

/** name of local host. */
char 		*hostname;

/** name of config file. */
char		conffile[256];

/** the table of storage methods */
const storage_method_t **storage_methods;

/** the current storage method */
const storage_method_t *storage_method;

/** the table of accounting agents */
const acc_agent_t **acc_agents;

/** the current accounting agent */
const acc_agent_t *acc_agent;

const access_agent_t **access_agents;
const access_agent_t *access_agent;


/** flag for output format. 0=user readable,
 *  1=for machine (ipacsum) */
int		machine_output_format = 0;
/** flag indicating if SIGALRM closed the databases */
int		db_closed_by_timer;

int in_transaction;

int storage_opened = 0;

int access_opened = 1;

char *storage_method_name = NULL;
char *acc_agent_name = DEFAULT_AGENT;
char *access_agent_name = DEFAULT_ACCESS;
char *authhost = NULL;
char *dbhost = NULL;
char *dbname = NULL;
char *dbport = NULL;
char *dbuser = NULL;
char *dbpass = NULL;
int slogin = 0;
char *ahost = NULL;

static const char usage_message_first[] =
"fetchipac Version " VERSION " - Usage:\n"
"fetchipac [options]\n"
"options are:\n"
"-S|--start\tinitial accounting setup: insert rules and chains\n"
"-C|--config FILE\t specify config file to use\n"
"-b|--batch\texecute commands which are fed via standard input\n"
"-H|--hostname\thostname to work with\n"
	"\t\tif not specified, fetchs with hostname() or hostname from config\n"
	"\t\treports are generated for specified host\n\t\t\tor for all hosts,"
	" if hostname is not specified\n"
"-f|--filter\tsimple filter. show only *<filter expresion>* rules\n"
"-d|--directory DIR\n"
	"\t\tspecify data directory DIR\n"
"-h|--help\tprint this help text\n"
"-m|--machine-output-format\n"
	"\t\toutput data in easily machine parsable format instead of\n"
	"\t\t easily human readable format\n"
"-r|--records\taccept timestamps on stdin, print corresponding records\n"
	"\t\t on stdout\n"
"-s|--storage-method METHOD\n"
	"\t\tuse METHOD as storage method. METHOD can be one of:\n"
	"\t\t ";
static const char usage_message_second[] = "\n"
	"\t\t Default is " DEFAULT_STORAGE "\n"
"-t|--timestamps [START][,END]\n"
	"\t\tlist timestamps. If given, start with timestamps START and\n"
	"\t\t end with timestamp END\n"
"-u|--sum {START][,END]\n"
	"\t\tprint summary for given timeframe\n"
"-R|--list-raw\tlist raw accounting rules\n"
"-v|--verbose\tadd one level of verbosity\n"
"See fetchipac(8) for further information.\n";

// check existence of chains and rules in kernel
// try to check whether rules in kernel are good
// return 0 in case of all ok
int
check_rules(void)
{
	acc_agent->init(0);
	return acc_agent->check();
}

int
list_raw(void)
{
	raw_rule_type *r;
	int i;

	if(access_agent->get_raw_list("", "", &r)) {
		fprintf(stderr, "access error\n");
		access_agent->close();
		return 1;
	}
	printf("rule name            | destination   | iface | proto | "
			"source              | destination         | direction | extensions\n");
	printf("------------------------------------------------------------------"
			"------------------------------------------------------\n");
	while(r) {
		printf("%-20s | %-13s | %-5s | %-5s | %-18s %s | %-18s %s | %-9s | %-8s",r->name,
			r->dest,r->iface,r->protocol,r->snet,r->sport,r->dnet,
			r->dport,r->direction,r->target);
		for (i=0;i<8;i++) {
			if (r->extension[i])
				printf(" %-8s ", r->extension[i]);
		}
		printf("\n");
		r = r->next;
	}
	return 0;
}

int
setup_access(int first)
{
	rule_type *tmp;

	if(acc_agent->init(0)) {
		perror("Failed to initialize accounting agent, "
			"aborting operations\n");
		return 1;
	}
	if(acc_agent->set(&tmp, first)) {
		perror("Failed to setup accounting agent, "
			"aborting operations\n");
		return 1;
	}
	return 0;
}

int
setup_agents(void)
{
	return setup_access(1);
}

/** print the usage message */
void
usage()
{
	const storage_method_t **p;

	printf(usage_message_first);
	for (p=storage_methods; *p != NULL; p++) {
		printf("%s%s", (*p)->name, (p[1] != NULL ? ", ": ""));
	}
	printf(usage_message_second);
}

/** set the hostname variable to the correct local host name. */
static void set_hostname()
{
	char hname[256];
	
	if (!hostname) {
		if (gethostname(hname, 256) != 0) {
			strcpy(hname, "(unknown host)");
			fprintf(stderr, "%s warning: cant get hostname, "
				"using '%s': %s\n",
				me, hname, strerror(errno));
		} else
			hostname = xstrdup(hname);
	}
}

/** create array storage_methods
 * @see storage_methods
 */
void
create_storage_methods()
{
	int i;
	const storage_method_t **p;
	const storage_method_t *(**pt)();

	for (i=0, pt=storage_method_if; *pt!=NULL; pt++, i++)
		;
	storage_methods = (const storage_method_t **)xmalloc((i+1) *
			sizeof(storage_method_t *));
	for (p = storage_methods, pt = storage_method_if; *pt!=NULL; p++, pt++)
		*p = (*pt)();
	*p = NULL;
}

/** create array acc_agents
 * @see acc_agents
 */
void create_acc_agents()
{
	int i;
	const acc_agent_t **p;
	const acc_agent_t *(**pt)();

	for (i=0, pt=acc_agent_if; *pt!=NULL; pt++, i++)
		;
	acc_agents = (const acc_agent_t **)xmalloc((i+1) *
			sizeof(acc_agent_t *));
	for (p = acc_agents, pt = acc_agent_if; *pt!=NULL; p++, pt++)
		*p = (*pt)();
	*p = NULL;
}

/** create array acc_agents
 * @see access_agents
 */
void create_access_agents()
{
	int i;
	const access_agent_t **p;
	const access_agent_t *(**pt)();

	for (i=0, pt=access_agent_if; *pt!=NULL; pt++, i++)
		;
	access_agents = (const access_agent_t **)xmalloc((i+1) *
			sizeof(access_agent_t *));
	for (p = access_agents, pt = access_agent_if; *pt!=NULL; p++, pt++)
		*p = (*pt)();
	*p = NULL;
}

void free_tree(void **ruletreep)
{
	/* The tree should be deleted here...
	 * Since fetchipac exists anyway, we don't NEED this.
	 */
}

/** make a nice ASCII string from a timestamp_t
 * s must be an array with at least 30 chars
 */
static void nice_time(timestamp_t t, char *s)
{
	struct tm *tmp;

	tmp = localtime(&t);

	strftime(s, 29, "%c", tmp);
}

int list_timestamps(timestamp_t tstart, timestamp_t tend, char *ahost)
{
	timestamp_t *tlist, *t, before, after;
	int n, i;
	char s[30];

	if (!storage_opened) {
		storage_method->open(SM_OPEN_READONLY);
		n = storage_method->list_timestamps(tstart, tend, &tlist,
			&before, &after, ahost);
		storage_method->close();
	} else
		n = storage_method->list_timestamps(tstart, tend, &tlist,
			&before, &after, ahost);
	
	if (n > 0) {
		printf(machine_output_format ? "%d\n"
				: "number of timestamps found: %d\n", n);
		if (before != (timestamp_t)-1) {
			if (machine_output_format)
				printf("- %lu\n", before);
			else {
				nice_time(before, s);
				printf("%12lu (%s) (final timestamp before "
						"start time)\n", before, s);
			}
		}
		for (i=0, t=tlist; i<n; i++, t++) {
			if (machine_output_format)
				printf("* %lu\n", *t);
			else
			{
				nice_time(*t, s);
				printf("%12lu (%s) ", *t, s);
				if (i%2 == 1)
					printf("\n");
			}
		}
		if (machine_output_format == 0)
			printf("\n");
		if (after != (timestamp_t)-1)
		{
			if (machine_output_format)
				printf("+ %lu\n", after);
			else
			{
				nice_time(after, s);
				printf("%12lu (%s) (first timestamp after "
						"end time)\n", after, s);
			}
		}
	} else
		printf(machine_output_format ? "0\n"
				: "number of timestamps found: 0\n");
	if (n > 0)
		free(tlist);
	return n>-1 ? 0 : 1;
}

/* print a record on file descriptor f */
int
print_records(FILE *f, int n, const data_record_type *dr)
{
	char buf[256];
	rule_type *r;
	int i;

	if (n < 0) {
		fprintf(f, "ERROR\n");
		return 1;
	}
	
	for (i=0; i<n; i++, dr++) {
		fprintf(f, "%s%lu ", machine_output_format ? "ADD\n":"timstamp: ",
			  dr->timestamp);
		if (machine_output_format == 0) {
			nice_time(dr->timestamp, buf);
			fprintf(f, "(%s) ", buf);
		}
		fprintf(f, "%s1\n", machine_output_format ? "":
			"number of records: ");
		if (machine_output_format == 0 && n>1)
			fprintf(f, "  record number: 1\n");
		fprintf(f, machine_output_format ?
				"( %s\n" : "    machine name: %s\n",
				/* FIXME: at the moment it's useless to
				          print the machine name returned
				dr->machine_name
				*/
				(ahost != NULL) ? ahost : "");
		for (r = dr->firstrule; r != NULL; r=r->next)
			fprintf(f, machine_output_format ?
				"%llu %llu |%s|\n"
				: "    bytes %14llu pkts %12llu %s\n",
				r->bytes, r->pkts, r->name);
		fprintf(f, "%s\n", machine_output_format ? ")" : "");
		printf("\n");
	}
	return 0;
}

/* list record data of timestamps on standard input */
int list_records(char *filter)
{
	timestamp_t timestamp_b, timestamp_e, *t, *tlist = NULL;
	data_record_type *dr = NULL;
	char buf[256], *cp;
	int n = 0, i, cnt;

	if (!storage_opened)
		storage_method->open(SM_OPEN_READONLY);

	while(fgets(buf, 256, stdin) != NULL) {
		cp = buf;
		if (*cp == '+' || *cp == '*' || *cp == '-')
			cp++;
		while (*cp!=0 && *cp==' ')
			cp++;
		timestamp_b = strtoul(cp, &cp, 0);
		if (*cp == '-') { //we also have an end timestamp
			cp++;
			timestamp_e = strtoul(cp, NULL, 10);
		}
		else
			timestamp_e = 0;

		if (timestamp_e 
			&& (!strcmp(storage_method->name, "postgre") ||
			    !strcmp(storage_method->name, "mysql")
			)) {
			/* for postgres/mysql storage and query for a timestamp range
			   do some memory optimizations */
			cnt = storage_method->list_timestamps(timestamp_b, timestamp_e, &tlist,
								NULL, NULL, NULL);
			
			for (i=0, t=tlist; i<cnt; i++, t++) {
				n = storage_method->get_records(*t, 0, &dr, filter);
				print_records(stdout, n, dr);

			}
			if (tlist) {
				free (tlist);
				tlist = NULL;
			}
		} else {
			/* all other storage methodes do it the default way */
			n = storage_method->get_records(timestamp_b, timestamp_e, &dr, filter);
			print_records(stdout, n, dr);
		}
		
		/* free memory
		   FIXME: in the future all backends should clear it's data at store_close() time */
		if (strcmp(storage_method->name, "postgre") 
		 && strcmp(storage_method->name, "sqlite") 
		 && strcmp(storage_method->name, "mysql") 
		 && (n>0))
			/* only for storage backends that don't already do free their own data 
			   we have to do it now */
			free_data_record_type_array(dr, n);
		
	}
	storage_method->close();
	fflush(stdout);
	return 0;
}

int print_summary(timestamp_t tstart, timestamp_t tend,
			char *ahost, char *filter)
{
	data_record_type *dr;
	int n;

	storage_method->open(SM_OPEN_READONLY);
	n = storage_method->get_summary(tstart, tend, &dr, filter);
	print_records(stdout, n, dr);

	if (n > 0)
		free_data_record_type_array(dr, n);

	storage_method->close();
	fflush(stdout);
	return 0;
}

/* spool a record. */
int spool_record(const data_record_type *dr)
{
	FILE *f;
	char *ahost_saved;
	int mof;
	int ret;

	f = fopen(spoolfile, "a");
	if (f == NULL) {
		fprintf(stderr, "%s: cant open spool file \"%s\" - data lost: "
				"%s\n", me, spoolfile, strerror(errno));
		return 1;
	}

	mof = machine_output_format;
	machine_output_format = 1;
	/* we have to use the ahost - see storage/postgre for details why
	   the hostname in the rules itself is ignored */ 
	ahost_saved = ahost;
	ahost = dr->machine_name;
	ret = 0;
	if (print_records(f, 1, dr) != 0) {
		fprintf(stderr, "%s: cant write data record to spool file \""
				"%s\", data lost: %s\n", me, spoolfile,
				strerror(errno));
		ret = 1;
	}
	fputs("\n", f);
	fclose(f);
	ahost = ahost_saved;
	machine_output_format = mof;
	return ret;
}

/* read spool file, write to database backend */
void unspool()
{
	FILE *f, *newf;
	struct stat stat_buf;
	int num;
	char line[256];		/* to read a line */
	char spoolfile_new[PATH_MAX];

	if (stat(spoolfile, &stat_buf) != 0)
		return;	/* file does not exist - very well */
	if (stat_buf.st_size == 0)
		return; /* file is empty - very well, too */

	f = fopen(spoolfile, "r+");
	if (f == NULL) {
		fprintf(stderr, "%s: cant open spool file \"%s\": %s\n",
				me, spoolfile, strerror(errno));
		return;
	}

	if (do_batchmode(f, &num) != 0) {
		/* error on adding records. */
		if (num > 0) {
			/* we need to rewrite the spool
			 * file, starting at the first failed record.
			 */
			if (fseek(f, 0, SEEK_SET) != 0) {
				fprintf(stderr, "%s: rewriting spool file: "
						"cant seek: %s\n", me,
						strerror(errno));
				/* what now... just leave the spool file
				 * untouched.
				 */
				fclose(f);
				return;
			}
			/* skip all records we dont want. */
			while(num > 0 && fgets(line, 256, f) != NULL) {
				if (strncasecmp(line, "ADD", 3) == 0)
					num--;
				if (strncasecmp(line, "BILL", 4) == 0)
					num--;
			}
			while(fgets(line, 256, f) != NULL) {
				if (strncasecmp(line, "ADD", 3) == 0)
					break;
				if (strncasecmp(line, "BILL", 4) == 0)
					break;
			}
			/* we now have the first line we want to keep */
			strcpy(spoolfile_new, spoolfile);
			strcat(spoolfile_new, ".new");
			newf = fopen(spoolfile_new, "w");
			if (newf == NULL) {
				fprintf(stderr, "%s: cant open new spool file "
						"\"%s\": %s\n", me,
						spoolfile_new,
						strerror(errno));
				fclose(f);
				return;
			}
			do {
				if (fputs(line, newf) == EOF) {
					fprintf(stderr, "%s: cant write to file"
						"\"%s\": %s\n", me,
						spoolfile_new,
						strerror(errno));
					fclose(newf);
					fclose(f);
					return;
				}
			} while(fgets(line, 256, f) != NULL);
			if (ferror(f)) {
				fprintf(stderr, "%s: error reading from spool "
						"file: %s", me,
						strerror(errno));
				fclose(f);
				fclose(newf);
				return;
			}
			fclose(f);
			fflush(newf);
			if (ferror(newf)) {
				fprintf(stderr, "%s: error writing to new spool"
						" file: %s", me,
						strerror(errno));
				fclose(newf);
				return;
			}
			fclose(newf);
			if (rename(spoolfile_new, spoolfile) != 0) {
				fprintf(stderr, "%s: error renaming \"%s\" to "
						"\"%s\": %s\n", me,
						spoolfile_new, spoolfile,
						strerror(errno));
			}
		}
	} else {
		if (unlink(spoolfile) != 0)
			fprintf(stderr, "%s: cant unlink \"%s\": %s - this "
					"might result in corrupted data\n",
					me, spoolfile, strerror(errno));
	}
}

/** enter batch mode. read commands from stdin and execute them.
 *  output results to stdout.
 *  valid commands are:
 *  quit	-	terminate
 * @return 0 if all commands executed okay, -1 in case of an error.
 *  errors for specific commands are reported on stdout.
 */
int batchmode()
{
	machine_output_format = 1;
	return do_batchmode(NULL, NULL);
}

/** fetchipac normally fetches data from the kernel. It can also list
 *  data from the database.
 */
int main(int argc, char **argv)
{
	rule_type *firstrule = NULL;

	int c;
	timestamp_t t, tstart=0, tend=0;
	data_record_type data_record;
	int mode;
	char *cp;
	const storage_method_t **smp;
	const acc_agent_t **aap;
	const access_agent_t **acp;
	int record_stored;		/* 1 if a record could be stored */
	char *filter=NULL;

	FILE *configa;

	static struct option long_options[] = {
		{ "agent",	    required_argument,	NULL,	'a'	},
		{ "config",	    required_argument,	NULL,	'C'	},
		{ "filter",	    required_argument,	NULL,	'f'	},
		{ "start",	    no_argument,	NULL,	'S'	},
		{ "list-raw",       no_argument,	NULL,	'R'	},
		{ "ahost",	    required_argument,  NULL,   'H'     },
		{ "batch",	    no_argument,	NULL,	'b'	},
		{ "directory",	    required_argument,	NULL,	'd'	},
		{ "help",	    no_argument,	NULL,	'h'	},
		{ "machine-output-format", no_argument,	NULL,	'm'	},
		{ "records",	    optional_argument,	NULL,	'r'	},
		{ "storage-method", required_argument,	NULL,	's'	},
		{ "timestamps",	    optional_argument,	NULL,	't'	},
		{ "verbose",	    no_argument,	NULL,	'v'	},
		{ "version",	    no_argument,	NULL,	'h'	},
		{ "sum",	    optional_argument,	NULL,	'u'	},
		{ NULL,		    0,			NULL,	0	}
	};

	me = argv[0];
	mode = 0;	/* fetch data */
	strncpy(conffile, CONFFILE, sizeof(conffile)-1); // default config

	/* create array storage_methods */
	create_storage_methods();
	create_acc_agents();
	create_access_agents();

	/* parse command line */
	while(1)
	{
		c = getopt_long(argc, argv, "a:L:O:C:f:SRUH:bd:vhmu::r::c:s:t::e",
				long_options, NULL);
		if (c==EOF)
			break;
		switch(c)
		{
			case 'a':
				access_agent_name = optarg;
				break;
			case 'b':
				mode = 3; /* batch mode */
				break;
			case 'S':
				mode = 5;
				break;
			case 'H':
				hostname = xstrdup(optarg);
				ahost = xstrdup(optarg);
				break;
			case 'f':
				filter = xstrdup(optarg);
				break;
			case 'C':
				strncpy(conffile, optarg, sizeof(conffile)-1);
				break;
			case 'R':
				mode = 4;
				break;
			case 'd':
				datadir = optarg;
				break;
			case 'h':
				usage();
				exit(1);
			case 'm':
				machine_output_format = 1;
				break;
			case 'v':
				verbose++;
				break;
			case 'r':
				mode = 2;
				break;
			case 's':
				storage_method_name = optarg;
				break;
			case 't':
				mode = 1; /* list timestamps */
				if (optarg != NULL) {
					tstart = strtoul(optarg, &cp, 0);
					if (*cp == ',')
						tend=strtoul(++cp, NULL, 0);
					else
						tend = time(NULL);
				} else {
					tstart = 0;
					tend = time(NULL);
				}
				break;
			case 'u':
				mode = 10; /* get summary */
				if (optarg != NULL) {
					tstart = strtoul(optarg, &cp, 0);
					if (*cp == ',')
						tend = strtoul(++cp, NULL, 0);
					else
						tend = time(NULL);
				} else {
					tstart = 0;
					tend = time(NULL);
				}
				break;
			case '?':
				fprintf(stderr, "%s: unknown option\n",	me);
				exit(1);
		}
	}

	configa = fopen(conffile, "r");
	if (configa != NULL) {
		parse_config(configa);
	} else {
		fprintf(stderr, "error opening config file '%s': %s\n", conffile, strerror (errno));
		return 2;
	}
	fclose(configa);
	set_hostname();

	/* find storage method */
	for (smp = storage_methods; *smp != NULL; smp++)
	{
		if (strcasecmp((*smp)->name, storage_method_name) == 0)
			break;
	}
	if (*smp == NULL) {
		fprintf(stderr, "%s: unknown storage method \"%s\" "
				"(not compiled in?)\n", me,storage_method_name);
		exit(1);
	}
	storage_method = *smp;

	/* find accounting agent */
	for (aap = acc_agents; *aap != NULL; aap++)
	{
		if (strcasecmp((*aap)->name, acc_agent_name) == 0)
			break;
	}

	if (*aap == NULL) {
		fprintf(stderr, "%s: unknown accounting agent \"%s\" "
				"(not compiled in?)\n", me,acc_agent_name);
		exit(1);
	}
	acc_agent = *aap;

	/* find access agent */
	for (acp = access_agents; *acp != NULL; acp++) {
		if (strcasecmp((*acp)->name, access_agent_name) == 0)
			break;
	}
	if (*acp == NULL) {
		fprintf(stderr, "%s: unknown access agent \"%s\" "
				"(not compiled in?)\n", me, access_agent_name);
		exit(1);
	}
	access_agent = *acp;


	if (access_agent->open(0)) {
		fprintf(stderr, "%s: error while opening access database\n", me);
		access_opened = 0;
		fprintf(stderr, "cant operate while database access is absent\n");
		return 1;
	}
	spoolfile = xmalloc(strlen(datadir) + sizeof(SPOOLFILE) + 10);
	sprintf(spoolfile, "%s/" SPOOLFILE, datadir);

	/* the normal mode is 0. If another mode is set, we do what has
	 * to be done here and leave main(). the normal fetch comes below
	 * this switch statement. you are not expected to like this
	 * program structure
	 */
	switch(mode)
	{
		case 1:
			return list_timestamps(tstart, tend, ahost);
		case 2:
			return list_records(filter);
		case 3:
			return batchmode();
		case 4:
			return list_raw();
		case 5:
			return setup_agents();
		case 10:
			return print_summary(tstart, tend, ahost, filter);
	}

// fetch data
	if (acc_agent->init(0)) {
		fprintf(stderr, "Error while initializing accounting agent\n");
		return 1;
	}
	if (check_rules()) {
		fprintf(stderr, "ipac-ng chains or rules corrupted, fix this "
				"with fetchipac -S\n");
		return 1;
	}

	if (lock(LOCKFILE) == 0) {
		struct stat *buf = NULL;

		buf=xcalloc(sizeof(struct stat), 1);
		storage_opened = 0;
		if (storage_method->open(0) != 0) {
			fprintf(stderr, "%s: storage reports error on open, "
					"switching to spool mode\n", me);
		} else
			storage_opened = 1;

		if (!stat(RECONFLAG, buf))
			if (buf->st_size>0) {
				mode=9;
				truncate(RECONFLAG, 0);
			}
		free(buf);
		time(&t);
		// FIXME:
		// Ugly hack to avoid some gdbm storage troubles :(
		// Will be removed together with gdbm storage
		// No, it will not be fixed soon %). only in 2.x :(
/*
		tmp = 1;
		while (tmp>0) {
			if (!storage_opened)
				tmp = 0;// storage closed? nothing to do, just spool
			else
				tmp = storage_method->list_timestamps(t, t, &tlist,
						&before, &after);
			if (tmp>0) t++;
		}
*/
		if (acc_agent->read(&firstrule) == 0) {
			data_record.timestamp = t;
			data_record.firstrule = firstrule;
			data_record.machine_name = hostname;
			record_stored = 0;

			if (dropzero) {
				rule_type *rule1 = firstrule;
				rule_type *rule2 = firstrule;
				
				while (rule1) {
					if (rule1->pkts==0)
						rule2->next=rule1->next;
					else
						rule2=rule1;
					rule1=rule1->next;
				}
			}
			
			if (storage_opened) {
				if (storage_method->store_record(&data_record) == 0)
					record_stored = 1;
				else
					fprintf(stderr, "%s: storage reports error on store record\n", me);
			}

			/* if our data could not be stored for some reason,
			 * we put it in ASCII representation into a spool
			 * file.
			 */
			if (record_stored == 0)
				spool_record(&data_record);
		}

		/* feed previously spooled records into the database. */
		if (storage_opened) {
			unspool();
			storage_method->close();
		}

		unlock(LOCKFILE);
	}
	if (access_opened)
		access_agent->close();
	return 0;
}
